### DISCUSSION TITLE
How to Use UUID in MySQL?

### DISCUSSION THREAD

**Discussion by RecordingTheSmile** [SeaQL/sea-orm#1561] [Q&A]
If I insert a active model with uuid like this:
```rust
UserEntity::ActiveModel {
    id:Set(my_uuid),
    data:Set(data)
}
```
I will get `UnpackInsertId` Err. So what should I do if i want to use uuid in mysql?

**Comment by billy1624** [comment]
Hey @whl2606555, please make sure your entity's primary key is defined with `auto_increment = false`. And upgrade to latest version of `sea-orm`

```rs
#[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel)]
#[sea_orm(table_name = "metadata")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    pub uuid: Uuid,
    // ...
}
```

**Comment by richard-fairthorne** [comment]
I stumbled across this discussion while using server generated UUID's with MySQL.

This simply does not work, because MySQL does not provide a facility for retrieving the LAST_INSERT_ID() of non-'auto_increment" columns.

This is different than the original problem that @whl2606555 posted, but I hope it helps anyone who is using server side functions to generate UUID primary keys, and stumbles across this post looking for a solution.

You simply cannot use server-generated primary keys, unless they are auto_increment.

**Reply by billy1624** [reply]
Hey @richard-fairthorne, I think you can define a create hook to auto-gen the UUID for new model.

```rs
#[async_trait]
impl ActiveModelBehavior for ActiveModel {
    /// Create a new ActiveModel with default values. Also used by `Default::default()`.
    fn new() -> Self {
        Self {
            uuid: Set(Uuid::new_v4()),
            ..ActiveModelTrait::default()
        }
    }

// ...
```

**Reply by richard-fairthorne** [reply]
> Hey @richard-fairthorne, I think you can define a create hook to auto-gen the UUID for new model.
> 

Absolutely. I tried this and it worked. The difficulty is that I am deriving my models from the db schema. I could not find a way to implement this outside of the automatically generated files. Each time the files are updated, the hooks are overwritten. 

I forget the details, but I think it's impossible to define a trait for a struct when neither the trait nor the struct are defined in your module.

**Reply by billy1624** [reply]
Implementing a trait for a struct on a different module is doable. What's can't is that implementing a trait for a struct on a different crate.

**Reply by richard-fairthorne** [reply]
> Implementing a trait for a struct on a different module is doable.

Yes. My recollection was poor. In fact the problem is that the derived models already contain a trait definition for the struct, and there cannot be two. 

I think I recall hearing about a technique where you wrap the struct in a new struct, and then create a new trait for the wrapper. It's not ideal, but may work ok.

**Comment by tyt2y3** [comment]
If you use UUID as PK in MySQL, generate it client side.
